### ISSUE TITLE
setProjectSettings() broken

### ISSUE THREAD

**Issue by grezniczek** [vanderbilt/redcap-external-modules#276]
It seems the (documented, framework **v1**) function `setProjectSettings()` is broken (lines 217ff in AbstractExternalModule.php; both _testing_ and in REDCap 9.7.7):

```php
/**
 * Saves all project settings (to be used with getProjectSettings).  Useful
 * for cases when you may create a custom config page or need to overwrite all
 * project settings for an external module.
 * @param array $settings Array of all project-specific settings
 * @param int|null $pid
 */
function setProjectSettings($settings, $pid = null)
{
	$pid = self::requireProjectId($pid);
	ExternalModules::saveSettings($this->PREFIX, $pid, json_encode($settings));
}
```

The **json_encode** is quite out of place, as `saveSettings() `expects an array as third argument. Thus, this leads to nothing being processed at all.

Which turns out to be a good thing, as the downstream processing of an actual array (like the one from `getProjectSettings()`, after removing json_encode in the snippet above) screws up the data stored in the _redcap_external_module_settings_ ... **badly**.

**Comment by mmcev106** [comment]
None of the modules on our instance reference this method.  I wonder if the best course of action would be to remove it from the docs, and make it email the system admin with a deprecation warning just in case anyone is actually using it.  If we don't hear of any reports of the email firing, we could remove the method entirely at some point in the future.

**Comment by 123andy** [comment]
I have vague memories of trying to use this and punting.  I do think it would be a valuable method for 'custom' configuration pages.  If we could make a simple single call to take the config params from an ajax post and dump them to the EM settings that would be convenient...

**Comment by grezniczek** [comment]
I guess nobody is actually using it as it simply does nothing.

But I just had another look at it. It would seem the issue with corruption I saw after removing json_encode and passing the data obtained with `getProjectSettings()` to `saveProjectSettings()` is caused by two things: 
1)  `getProjectSettings()` is calling `ExternalModules::getProjectSettingsAsArray()` without specifying `false` as the third argument, thus returning project **AND** system settings. 
2) The raw settings that are expected by `ExternalModules::saveSettings()` are structured totally different from what getProjectSettings() returns, or even what `ExternalModules::formatRawSettings()` transforms them into.

It would seem the solution is to transform the result from `getProjectSettings()`  (which we probably cannot change now) and which, as per docs, is supposed to be the input (after modification) to `setProjectsSettings()`, into something that matches the output of `ExternalModules::formatRawSettings()`.

I was trying to put this into a PR, but came accross the issue, that `getProjectSettings()` puts copies of system-only values into the project `value` key. This leads values that are supposed to be system-only (and may represent sensitive information) being copied (and then stored) as project values, which exposes them in the next config call to anybody with project design rights.

So to fix this, `getProjectSettings()` needs to be fixed, and this can only be done with a **breaking change** (i.e. having it return project settings only, no more system settings, and then probably best without the `value` indirection.

**Should we do that?**

**Comment by mmcev106** [comment]
Unless I end up working on a project that needs this method to work properly, the only thing I can justify doing is deprecating it.

**Comment by grezniczek** [comment]
This would still leave the issue of `getProjectSettings()` returning copies of system values as if they were project settings, which may lead to inadvertently exposing sensitive information if module authors are not aware of this (happily sending of everything to the browser). And this could even lead to such values being stored as project values when someone decides to iterate over the array returned by `getProjectSettings()` to save values with `setProjectSetting()`.

In my opinion, this alone warrants addressing. And once this is fixed, the fix for `setProjectSettings()` is basically free.

**Comment by mmcev106** [comment]
PRs are welcome.  I could also justify removing the method entirely if there's a very low possibility that anyone is actually using it.

**Comment by grezniczek** [comment]
@mmcev106, @123andy  ... I have issued a PR with a fix (effective with framework v4). #277

**Comment by 123andy** [comment]
I've never liked the concept of getting a project setting only to receive a system setting if not set.   It seems to cause more potential problems then gains.   If I want to have the same setting on a system and project level, I've just checked both levels in my code.  So, I think I'm supportive if I follow correctly...

**Comment by mmcev106** [comment]
I regret that design in hindsight as well...  I'm hopefully I'm be able to review this within the week. Assuming there are no cases where it could break existing functionality, I'll go ahead and merge it.

**Comment by grezniczek** [comment]
The PR does not change existing functionalities for framework versions 1-3.
Version 4 will be a breaking change for `getProjectSettings()` as the return format differs.

**Comment by grezniczek** [comment]
I see that Framework v4 will be included in the upcoming 9.7.8. I guess this will mean a fix for this issue here will only be possible whenever a v5 comes along?

**Comment by mmcev106** [comment]
Yes, I believe it makes sense to introduce this in v5.  I'm sorry, I'm not sure when I'll have an opportunity to consider this and other outstanding PRs at the moment.

**Comment by mmcev106** [comment]
I think it makes sense to introduce v5 just for this (or any breaking change).

**Comment by grezniczek** [comment]
Ok, I will update the PR accordingly.

**Comment by mmcev106** [comment]
Thanks to @grezniczek, this has been fixed via PR #277!  Closing...
